lib/streamlit/elements/file_uploader.py

```diff
@@ -187,6 +192,22 @@ def serialize_file_uploader(
             serializer=serialize_file_uploader,
         )
 
+        ctx = get_report_ctx()
+        if ctx is not None and widget_value:
+            serialized = serialize_file_uploader(widget_value)
+
+            # The first number in the serialized widget_value list is the id
+            # of the most recently uploaded file.
+            newest_file_id = serialized[0]
+            active_file_ids = list(serialized[1:])
+
+            ctx.uploaded_file_mgr.remove_orphaned_files(
+                session_id=ctx.session_id,
+                widget_id=file_uploader_proto.id,
+                newest_file_id=newest_file_id,
+                active_file_ids=active_file_ids,
+            )
```

Comment:
If the file_uploader is removed from the script, will its orphaned files ever be collected? Not something to address in this PR (and maybe I'm misunderstanding), but I wonder if this behavior should eventually be changed to occur for all a session's file_uploaders at the end of each script run.

Reply:
Hm, yeah I think you're right that if a `file_uploader` widget is deleted, its orphaned files will never be GCd. I'll file an issue to revisit this later as I believe the issue existed pre-session_state